include/midi_pedal.hpp

```diff
@@ -7,26 +7,34 @@ typedef void (*note_off_callback_t)(uint8_t, uint8_t);
 class MidiPedalAdapter {
   public:
 	MidiPedalAdapter(note_on_callback_t note_on_callback, note_off_callback_t note_off_callback)
-	    : held_lo{0}, held_hi{0}, pedal_down{0}, note_on_callback(note_on_callback),
+	    : note_held{0}, note_sustained{0}, pedal_down{0}, note_on_callback(note_on_callback),
 	      note_off_callback(note_off_callback) {}
 
 	void set_pedal(uint8_t channel, bool pedal_down) {
-		if (this->pedal_down[channel] && !pedal_down) {
-			for (int i = 0; i < 64; i++) {
-				if (held_lo[channel] & (1 << i)) {
+
```

Comment:
This completely ignores the `channel` argument now. Doing so is almost always a reason to think about it twice:

- Either you just made a mistake by violating the contract your function gave to the "outside world": It told everyone that it takes a `channel` argument and does something sensible with it. But in fact, it just ignores it. Likely this is not right.
- Or you want to actually change the API that is announced publically, by removing that argument altogether.

However think about this: What it I press the pedal, play some notes on channel 1, and *release* the same notes on channel 2. They would get stuck, because whatever listens on channel 1 still thinks they're pressed, but the pedal adapter will think that it has already released them (and not release them again).

So i firmly believe that we'll actually need to keep 16 arrays of 128 notes each.

Try it with the following code:

```
/** Acts very similar to a bool[8], but differs from it in two things:
  * Firstly, it stores the bools more efficiently.
  * Secondly, you have to use my_array.get(4) and .set(4, true) instead of myarray[4].
  */
struct BooleanArray8 {
    void set(uint8_t index, bool value) {
        value = value & (~(1 << index)) | ((!!value) << index));
    }
    bool get(uint8_t index) {
        return value & (1 << index);
    }

    private:
        uint8_t value = 0;
}

/** Acts very similar to a bool[128], but differs from it in two things:
  * Firstly, it stores the bools more efficiently.
  * Secondly, you have to use my_array.get(42) and .set(42, true) instead of myarray[42].
  */
struct BooleanArray128 {
    void set(uint8_t index, bool value) {
        int subarray_index = index / 8;
        int bit_in_subarray = index % 8;
        subarray[subarray_index].set(bit_in_subarray, value);
    }
    void get(...) { /* should be obvious now */ }

    private:
        BooleanArray8 subarray[16]; // 16 * 8 = 128.
}
        
```

Reply:
Cool Thanks for the suggestion.. first line of struct BooleanArray8 {void set(uint8_t index, bool value) { was throwing an error (parenthisis)
Is this fix correct?
    
`value = (value & (~(1 << index)) )| (((!!value) << index));`

Reply:
I tried implementing the strucs instead of the bool arrays. Code compiles but the strucs .get appear to always return true (the pedal sticks again)... I pushed the latest code. Any thoughts? Thanks

Reply:
Now that it is optimized, what is the recommended strategy to make the pedal adaptor multi channel?

Perhaps make the pedal adaptor monophonic (removing channel stuff) and declare 4 instances of it in midi.cpp?
TherapSid only handles 4 channels so might be better?

Thanks!!!

Reply:
Rather than having only one (two) of these `BooleanArray128` (which suffices for only one channel), you need 16 (2*16) of them.

You can have a `BooleanArray128 heldNotes[16];` and the same for `sustainedNotes`.

`note_on` and `note_off` would access them based on the `channel`, and for `set_pedal` you have two options:

- either, you keep the `channel` argument in `set_pedal` and make it respect the channel, too. In this case you would want to `for (int c=0; c<16; c++) blah.set_pedal(c, actual_pedal_value)` in `midi.cpp` if you still want channel 1's pedal to control *all* the channels. I'd prefer that solution because it makes the `MidiPedalAdapter` more reusable and less surprising in its behavior. (this requires a `pedal_down[16]` array, too.)
- or, you remove the `channel` argument in `set_pedal`. in this case, `set_pedal` needs to loop over all 16 channels.

Reply:
Sounds good, I'll take the suggested BooleanArray128 heldNotes[16]; sustainedNotes[16]; approach, thanks!

Reply:
Should be sorted now, I'll let you resolve if you're happy with it :)